### PR TITLE
Add Stream payload to ExecuteStoredProcedureStreamAsync 

### LINKED ISSUE #1059: ExecuteStoredProcedureStreamAsync is only half-Stream
tldr: I would like an overload of `ExecuteStoredProcedureStreamAsync` which takes a `Stream` as a parameter.

Scripts.`ExecuteStoredProcedureStreamAsync` returns a ResponseMessage (aka Stream) so we can handle it however we wish.  Streams are handy because we can control serialization (and other reasons).

However, it takes `dynamic[]` as a parameter representing the parameters to the stored procedure.  In ScriptsCore, this is converted to a stream using `this.clientContext.CosmosSerializer.ToStream<dynamic[]>(parameters);`.  In other words, it uses CosmosSerializer.  

While I suppose this is fine, I would like more control over when things get serialized.

From a higher-level point of view, I would prefer that my entire experience with Cosmos is Stream-based, and that when using Stream methods, there's "no such thing" as serialization.  In other words, Stream-based methods should never serialize nor deserialize "user" objects (but can still serialize internal structures, eg: change feed [documents:[...]]).

Challenges with adding an overload to ExecuteStoredProcedureStreamAsync:
- The system expects an array but the user passes something else, and as a stream, we can't tell the difference until the server probably gives us a BadRequest response.
- `dynamic[] parameters` and `Stream streamPayload` cannot be in the same position because `null` can be passed in and this would be a breaking change

Mitigations:

If the user serializes something that's not an array, just allow it and let the server blow up with a BadRequest.  No big deal.

As to the parameter positions:
Option 1:  Change the positions.  It seems "awkward" but we can do `string storedProcedureId, Stream streamPayload, PartitionKey partitionKey` which is safely differentiated from `string storedProcedureId, PartitionKey partitionKey, dynamic[] parameters`
Option 2: Change the name.  But I can't think of a non-awkward alternative name for `ExecuteStoredProcedureStreamAsync`.

I'll post a PR for this.  But I realize I might be stepping on some design-toes so feel free to consider the PR as just a further explanation to this issue.

### PR DESCRIPTION
**Pull request by joshlang**
## Description

#1059 

This adds an overload to ExecuteStoredProcedureStreamAsync which accepts a `Stream` payload representing the parameters.  The original only accepted `dynamic[]`, which must be serialized.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Closing issues

closes #1059

### CODE REVIEW

**Comment by joshlang** [comment]
Worth a shot :D

**Comment by joshlang** [comment]
Wow, you guys have quite the test suite.  I've updated my local environment so that I can actually run them all.  Previously, most failed so I just ignored failures.

FYI Microsoft.Azure.Cosmos.Tests -> AsyncCacheTest -> TestCancelOnOneThreadCancelsOtherTaskIfCanceled still fails for me.

**Comment by joshlang** [comment]
Unrelated to this PR - just a question:

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/94d517b69245fd75185d8e364c721cb3fe42da74/Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/AsyncCacheTest.cs#L168

This assertion is what fails the test when I run it locally.  

The assertion doesn't seem to make sense - it's a race condition at best.  After the cancellationTokenSource is .Cancel()'ed, getTask1 can complete at any time, including immediately.

What's your approach to stuff like this?  Leave it alone unless it's actually causing trouble in the pipeline?  Raise an issue and let the test owner fix it?  Hope for a PR lol?

**Comment by j82w** [comment]
Create an issue for the failing test. Someone will investigate it. It probably won't be until next week. What's interesting is the test seems to be passing in the gates. Please make sure to put the OS you are running.

/azp run

**Comment by joshlang** [comment]
Test failed due to ...`Class Initialization method Microsoft.Azure.Cosmos.SDK.EmulatorTests.BatchSinglePartitionKeyTests.ClassInitialize threw exception. Microsoft.Azure.Cosmos.CosmosException: Microsoft.Azure.Cosmos.CosmosException: Response status code does not indicate success: 408 Substatus: 0 Reason: (Microsoft.Azure.Documents.RequestTimeoutException`...

Probably transient.

**Comment by joshlang** [comment]
Could I get a friendly neighborhood /azp run por favor?

**Comment by kirankumarkolli** [comment]
#1059 updated with more context and follow-up.

**Comment by joshlang** [comment]
changelog.md conflicts are now resolved.

**Comment by kirankumarkolli** [comment]
@ealsur can you please take a look a the tests and any new coverage needed?

**Comment by joshlang** [comment]
Failing test - probably transient:
```
X TestReducePageSizeScenario [393ms]
  Error Message:
   Initialization method Microsoft.Azure.Cosmos.SDK.EmulatorTests.ChangeFeed.DynamicTests.TestInitialize threw exception. Microsoft.Azure.Cosmos.CosmosException: Response status code does not indicate success: 500 Substatus: 0 Reason: (Microsoft.Azure.Documents.DocumentClientException: Message: {"Errors":["An unknown error occurred while processing this request."]}
...
```

File: Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs [hunk 1]
```diff
@@ -416,6 +417,73 @@ public abstract class Scripts
             StoredProcedureRequestOptions requestOptions = null,
             CancellationToken cancellationToken = default(CancellationToken));
 
+        /// <summary>
+        /// Executes a stored procedure against a container as an asynchronous operation in the Azure Cosmos service and obtains a Stream as response.
+        /// </summary>
+        /// <param name="storedProcedureId">The identifier of the Stored Procedure to execute.</param>
+        /// <param name="streamPayload">(Optional) An optional <see cref="Stream"/> containing the payload, which should represent an array of parameters for the stored procedure.</param>
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs]: If the Stream is optional, it should be after all required parameters and have a `= null` default value?

File: Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs [hunk 2]
```diff
@@ -416,6 +417,73 @@ public abstract class Scripts
...
+        /// <param name="partitionKey">The partition key for the item.</param>
+        /// <param name="requestOptions">(Optional) The options for the stored procedure request.</param>
+        /// <param name="cancellationToken">(Optional) <see cref="CancellationToken"/> representing request cancellation.</param>
+        /// <returns>The task object representing the service response for the asynchronous operation which would contain any response set in the stored procedure.</returns>
+        /// <exception cref="ArgumentNullException">If <paramref name="storedProcedureId"/> or <paramref name="partitionKey"/>  are not set.</exception>
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs]: Would be good to document what is the expected StatusCode / Response if the Stream does not represent an array of parameters

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/StoredProcedureTests.cs [hunk 3]
```diff
@@ -436,6 +539,46 @@ public async Task ExecuteTestWithMultipleParameters()
             Assert.AreEqual(HttpStatusCode.NoContent, deleteResponse.StatusCode);
         }
 
+        [TestMethod]
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/StoredProcedureTests.cs]: Add a test to understand what happens if the Stream is not an array? That way we can catch changes in the backend behavior and assert we don't accidentally change the expected output in a future change

File: Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs (hunk 1 above)
Comment (joshlang) [reply Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs]: This is a difficult one because of the challenges with positional parameters here: https://github.com/Azure/azure-cosmos-dotnet-v3/issues/1059

We are competing with this overload:
```
string storedProcedureId, PartitionKey partitionKey, dynamic[] parameters, StoredProcedureRequestOptions requestOptions = null, CancellationToken cancellationToken = default(CancellationToken)
`````
And ours is: 
```
string storedProcedureId, Stream streamPayload, PartitionKey partitionKey, StoredProcedureRequestOptions requestOptions = null, CancellationToken cancellationToken = default(CancellationToken)
```

First, I would point out that `dynamic[] parameters` is optional, but does not have a default value of null.  This makes sense because `null` is not by default a valid parameter - it only is if the stored procedure requires no parameters.  It makes sense to me that the user should explicitly specify `null` if their scenario requires it.

My recommendation in this would be to leave it as is, and not to provide any default parameter for either overload.

We could update the documentation (for both overloads) to be more clear, though?  Perhaps more accurate than *optional* would be to say that *null is acceptable*.  

Examples:
`(Optional) An array of dynamic objects representing the parameters for the stored procedure.`
becomes
`An array of dynamic objects representing the parameters for the stored procedure, or optionally null if there are no parameters.`

and

`(Optional) An optional <see cref="Stream"/> containing the payload, which should represent an array of parameters for the stored procedure.`
becomes
`A <see cref="Stream"/> containing the payload which should represent an array of parameters for the stored procedure, or optionally null if there are no parameters.`

...

If that is not acceptable to you, I'll think about alternatives for either method naming or parameter placement.  The challenge is not to break existing code where `null` is being passed and now is ambiguous between `dynamic[] parameters = null` and `Stream streamPayload = null`.

File: Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs (hunk 2 above)
Comment (joshlang) [reply Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs]: I have updated the `<returns>` section.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/StoredProcedureTests.cs (hunk 3 above)
Comment (joshlang) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/StoredProcedureTests.cs]: GOOD CALL!

Doing so uncovered some quirkiness.

I found some weird edge cases.  It was mysterious until I outputted `arguments.callee.caller.toString()` and found:
```
function __docDbMain() { 
  var fn = function(param1, param2) { my stored proc };
  ...
  var args = JSON.parse(...my stream payload...);
  var ret = fn.apply(fn, args);
  return ret;
}
```

It turns out that cosmos will accept:

- An array
- Any valid object
- The null literal
- null

Cosmos return (400) BadRequest when passed:

- Invalid JSON
- Non-objects (number, boolean, string, undefined)

When an object is passed, if it is array-like, cosmos handles it accordingly, leading to:

- `{ "0": "hello", length: 1 }` passes one 'hello' parameter
- `{ "0": "hello", "1": "there", length: 5 }` passes one 'hello' parameter (not 'there')
- `{ "banana": [ "bread" ] }` is ignored entirely
- `{ "0": "hello", length: 60000 }` causes a stack overflow
- `{ "0": "hello", length: 1e9 }` fails because .apply rejects it

So, I have added test cases to cover all these.  

I've also changed the documentation.  It was difficult to be succinct.  The most succinct way to explain it was to reference JSON.parse and Function.apply.  Hopefully that's acceptable.

The `<param>` element now reads: `A <see cref="Stream"/> containing the payload which should represent a JSON array or arraylike object of parameters. This is parsed using JSON.parse and Function.apply uses the result to call the stored procedure. This can be null if no parameters are required.`

Normally, I'd try and pave over some of these edge cases and hide them from the user.  I'm especially nervous about referring to the internal workings of cosmos (ie JSON.parse, fn.apply) because they may change.

That being said, this functionality is already public.  It's out there.  Even if the internal workings do change, it's the cosmos service team that's making the breaking change.  So I still do not have qualms about exposing this `streamPayload` overload.

File: Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs (hunk 1 above)
Comment (joshlang) [reply Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs]: Alright, the commits I've made resolved the other two conversations.  I've also committed proposed changes to resolve your concerns raised in this conversation.

I'd need your feedback/guidance before considering this resolved.  Or if everything's great, then cool :D

**Comment by joshlang** [comment]
> /asp run

azp? :D

File: Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Scripts/Scripts.cs]: I see, the `dynamic` on the other signature makes it tricky. Since @kirankumarkolli signed off, I guess it's ok 😄

**Comment by ealsur** [comment]
There seems to be a build error from the merge with master. `ScriptsInlineCore` needs to implement the new public method, it should be similar to the other Execute where TaskHelper is used to wrap the inner call (ref #1078)

**Comment by joshlang** [comment]
You silly goose, why do you update master behind my back :D 

Ok, will investigate.

**Comment by ealsur** [comment]
Sorry! It should be as simple as:

    public override Task<ResponseMessage> ExecuteStoredProcedureStreamAsync(
        string storedProcedureId,
        Stream streamPayload,
        PartitionKey partitionKey,
        StoredProcedureRequestOptions requestOptions = null,
        CancellationToken cancellationToken = default(CancellationToken));
    {
        return TaskHelper.RunInlineIfNeededAsync(() => this.scripts.ExecuteStoredProcedureStreamAsync<TOutput>(storedProcedureId, streamPayload, partitionKey, requestOptions, cancellationToken));
    }

**Comment by joshlang** [comment]
No problem.  Synced up from Azure/master, made the changes and should be good to go

**Comment by joshlang** [comment]
Various connectivity errors to nuget:
```
C:\Program Files\dotnet\sdk\3.1.100\NuGet.targets(123,5): error : Unable to load the service index for source https://api.nuget.org/v3/index.json. [D:\a\1\s\Microsoft.Azure.Cosmos.sln]


Failed to download package 'Microsoft.Azure.Cosmos.Serialization.HybridRow.1.0.0-preview' from 'https://www.myget.org/F/azure-cosmos/api/v3/flatcontainer/microsoft.azure.cosmos.serialization.hybridrow/1.0.0-preview/microsoft.azure.cosmos.serialization.hybridrow.1.0.0-preview.nupkg'.
         No connection could be made because the target machine actively refused it.
```

Either nuget has exploded permanently, or hopefully it's transient
